src/Resolver.php

```diff
@@ -598,14 +599,18 @@ public function requireAndResolveColumns(array $columns, Model $model = null)
 
             if (! $column instanceof ExpressionInterface) {
                 $column = $this->getBehaviors($target)->rewriteColumn($column, $relationPath) ?: $column;
-            }
 
-            if (
-                ! $column instanceof ExpressionInterface
-                && ! $this->hasSelectableColumn($target, $columnPath)
-                && ! $this->hasSelectableColumn($target, $alias)
-            ) {
-                throw new InvalidColumnException($columnPath, $target);
+                if (
+                    ! $this->hasSelectableColumn($target, $columnPath)
+                    && ! $this->hasSelectableColumn($target, $alias)
+                ) {
+                    throw new InvalidColumnException($columnPath, $target);
+                }
+
+                if (isset($targetColumns[$column])) {
+                    $alias = is_int($alias) ? $column : $alias;
+                    $column = $targetColumns[$column];
```

Comment:
If this is an expression afterwards, it needs to be resolved still. Just like it is being done at the start of the foreach. That's why the tests fail as well.